File: frontend/src/api/messages.ts
import { api } from "../lib/api";
import type { Message, MessageCreate, FollowUpRequest, FollowUpResponse } from "../types";

export const messagesApi = {
  /**
   * Get all messages for a case (conversation thread)
   */
  async list(caseId: string): Promise<Message[]> {
    const response = await api.get(`/cases/${caseId}/messages`);
    return response.data;
  },

  /**
   * Create a new user message (follow-up question)
   * @deprecated Use followUp() for lightweight conversational follow-ups
   */
  async create(caseId: string, data: MessageCreate): Promise<Message> {
    const response = await api.post(`/cases/${caseId}/messages`, data);
    return response.data;
  },

  /**
   * Submit a follow-up question for async processing (HTTP 202 Accepted).
   *
   * Returns immediately with the user message. The assistant response is
   * generated in the background. Poll case status until "completed" to
   * get the assistant response via list().
   *
   * Flow:
   * 1. Submit follow-up → get back user message (202 Accepted)
   * 2. Case status changes to "processing"
   * 3. Background task generates assistant response
   * 4. Case status changes to "completed"
   * 5. Call list() to get all messages including assistant response
   */
  async followUp(caseId: string, data: FollowUpRequest): Promise<FollowUpResponse> {
    const response = await api.post(`/cases/${caseId}/follow-up`, data);
    return response.data;
  },
};

File: backend/api/messages.py
"""Message API endpoints for conversation threading."""

from typing import List
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from api.schemas import MessageCreate, MessageResponse
from api.middleware.auth import get_current_user
from db.connection import get_db
from db.repositories.message_repository import MessageRepository
from db.repositories.case_repository import CaseRepository
from models.user import User

router = APIRouter()


@router.get("/cases/{case_id}/messages", response_model=List[MessageResponse])
def get_case_messages(
    case_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Get all messages for a case, ordered chronologically.

    Returns the conversation thread including both user questions
    and assistant responses.
    """
    # Verify case exists and user has access
    case_repo = CaseRepository(db)
    case = case_repo.get(case_id)
    if not case:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Case {case_id} not found"
        )

    if case.user_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You don't have access to this case"
        )

    # Get all messages for the case
    message_repo = MessageRepository(db)
    messages = message_repo.get_by_case(case_id)

    return messages


@router.post("/cases/{case_id}/messages", response_model=MessageResponse, status_code=status.HTTP_201_CREATED)
def create_message(
    case_id: str,
    message: MessageCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Create a new user message (follow-up question) for a case.

    This adds a message to the conversation thread. The frontend
    should then trigger analysis to generate an assistant response.
    """
    # Verify case exists and user has access
    case_repo = CaseRepository(db)
    case = case_repo.get(case_id)
    if not case:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Case {case_id} not found"
        )

    if case.user_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You don't have access to this case"
        )

    # Create user message
    message_repo = MessageRepository(db)
    new_message = message_repo.create_user_message(
        case_id=case_id,
        content=message.content
    )

    return new_message
